### PR TITLE
ASoC: SOF: Intel: Harden DSP detection with class/subclass/prog-if

### PR DESCRIPTION
**Pull request by plbossart**
The use of PPCAP and GCAP to detect the DSP presence is not reliable
or meaningful. The recommendation is to use the class/subclass/prog-if
information instead (as done by Windows drivers)

Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>

### CODE REVIEW

**Comment by plbossart** [comment]
Reviewers, can you please test on your development platforms? The DSP detection is based on information shared by Windows folks, I only tested this code on SKL and WHL and I'd like to make sure it's functional across the board. I need APL, GLK, WHL, CNL, ICL results, thanks for your help!

**Comment by bardliao** [comment]
@plbossart Both Up2 (APL) and ICL show "sof-audio sof-audio: DSP detected with PCI class/subclass/prog-if 0x040100".

**Comment by RanderWang** [comment]
I only have WHL, it is: DSP detected with PCI class/subclass/prog-if 0x040100
WHL is alias of CNL, So CNL should be like this.

**Comment by ranj063** [comment]
@plbossart verified on GLK.
[    7.733835] sof-audio sof-audio: DSP detected with PCI class/subclass/prog-if 0x040100
[    7.733973] sof-audio sof-audio: found DSP capability at 0x800
[    7.768422] sof-audio-pci 0000:00:0e.0: PP capability, will probe DSP later.

File: sound/soc/sof/intel/hda.c [hunk 1]
```diff
@@ -524,6 +524,22 @@ int hda_dsp_probe(struct snd_sof_dev *sdev)
...
+	 * class=04 subclass 03 prog-if 00: no DSP, legacy driver is required
+	 * class=04 subclass 01 prog-if 00: DSP is present
+	 *   (and may be required e.g. for DMIC or SSP support)
+	 * class=04 subclass 03 prog-if 80: either of DSP or legacy mode works
+	 */
+	if (pci->class == 0x040300) {
+		dev_err(sdev->dev, "The DSP is not enabled on this platform, aborting probe\n");
+		return -ENODEV;
+	} else if (pci->class != 0x040100 && pci->class != 0x040380) {
+		dev_err(sdev->dev, "Unknown PCI class/subclass/prog-if 0x%06x found, aborting probe\n", pci->class);
```
Comment (bardliao) [inline sound/soc/sof/intel/hda.c:538]: Here is over 80 characters. So, checkpatch.pl will complain about it

**Comment by plbossart** [comment]
On 12/6/18 2:04 AM, Bard Liao wrote:
> *@bardliao* commented on this pull request.
> 
> checkpatch.pl wiil complain about "line over 80 characters"

Yes, but I will tolerate lines over 80 characters for error messages.
Likewise if there is an extra line after a for or while loop that's ok 
by me.

**Comment by plbossart** [comment]
Great, thanks all for your tests. It does look like a reliable indicator so merging now.

**Pull request merged**